### PR TITLE
Add dead letter blob processing

### PR DESCRIPTION
**Pull request by kkuzmin**
### Problem Description
Collect function can fail during data processing and can now produce some dead letters stored in an Azure blob.

### Solution Description
Add basic functionality for dead letter blob processing which helps to simplify `DLBlob` function development on a collector side.

### CODE REVIEW

File: dlblob.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            },
+            function(blobData, blobReq, blobResp, asyncCallback) {
+                try {
+                    return dlblob._processCallback(dlblob._context, blob, blobData, asyncCallback);
+                } catch (ex) {
+                    return asyncCallback(ex);
+                }
+            },
+            function(asyncCallback) {
+                return dlblob._blobService.deleteBlob(dlblob._dlContainerName, blob.name, asyncCallback);
```
Comment (mtrybulec) [inline dlblob.js]: In the original ehub collector, result.skipped === 0 was checked before deleting the blob - is that no longer needed?

File: test/dlblob_test.js [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        .reply(202, function() {deleteBlobStub();});
+        
+        var testProcessingStub = sinon.stub().callsFake(
+            function(context, dlblob, dlblobPayload, callback) {
+                sinon.assert.match(dlblob.name, 'kktestdl/ehubgeneral/2019-01-23T15');
+                assert.equal(context.executionContext.invocationId, 'invocation-id');
+                return callback(null);
+            });
+        var dlblob = new AlAzureDlBlob(mock.DEFAULT_FUNCTION_CONTEXT, testProcessingStub);
+        dlblob.processDlBlobs({}, function(err) {
```
Comment (mtrybulec) [inline test/dlblob_test.js]: Perhaps check that err is null?

File: test/dlblob_test.js [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        
+        var testProcessingStub = sinon.stub().callsFake(
+                function(context, dlblob, dlblobPayload, callback) {
+                    return callback(null);
+                });
+        var dlblob = new AlAzureDlBlob(mock.DEFAULT_FUNCTION_CONTEXT, testProcessingStub);
+        dlblob.processDlBlobs({}, function(err) {
+            assert.equal(err.code, 'ContainerNotFound');
+            assert.equal(err.statusCode, 404);
+            sinon.assert.callCount(testProcessingStub, 1);
```
Comment (mtrybulec) [inline test/dlblob_test.js]: How many ties will deleteBlob be called here, 6? Perhaps worth adding.

File: test/dlblob_test.js [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        nock('https://kktestdl.blob.core.windows.net:443', {'encodedQueryParams':true})
+        .get(/alertlogic-dl.*/)
+        .times(6)
+        .reply(200, mock.GET_BLOB_CONTENT_TEXT);
+        
+        // Delete blob
+        nock('https://kktestdl.blob.core.windows.net:443', {'encodedQueryParams':true})
+        .delete(/alertlogic-dl\/kktestdl\/ehubgeneral.*/)
+        .times(6)
+        .reply(404, mock.CONTAINER_NOT_FOUND);
```
Comment (mtrybulec) [inline test/dlblob_test.js]: Perhaps return 404 on a 2nd or later try, to show that a failed deleteBlob doesn't interrupt blob processing (that is, testProcessingStub call count is still 6).

File: dlblob.js (hunk 1 above)
Comment (kkuzmin) [reply dlblob.js]: Good point. I'll make sure this check still present in `ehub` DL processing callback.
Here in the library we check if processing callback  return `null` then we delete the blob.

File: test/dlblob_test.js (hunk 3 above)
Comment (kkuzmin) [reply test/dlblob_test.js]: Good catch. It should be called ones. And it is supposed to be checked here https://github.com/alertlogic/al-azure-collector-js/pull/13/files#diff-ef57d0ffdfcda235d1b919f3630bd9fcR142

I'll change it to 1 time.

**Comment by kkuzmin** [comment]
@mtrybulec  thank you for the comments. I reworked DL blob processing. Now it isn't stopped if processing fails for a just one blob.
Basically every DL blob processing happens independently from each other.

**Pull request merged**